src/Castle.Core.Tests/DynamicProxy.Tests/InParamsTestCase.cs

```diff
@@ -0,0 +1,165 @@
+			proxy.Method(in x);
+
+			Assert.AreEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_reference_Ref_parameter_can_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IByRef>(new SetArgumentValueInterceptor(different));
+			proxy.Method(ref x);
+
+			Assert.AreNotEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		public void By_reference_Out_parameter_can_modify_argument_var()
+		{
+			var x = new ReadOnlyStruct(1);
+			var original = new ReadOnlyStruct(x.Value);
+
+			var different = new ReadOnlyStruct(x.Value + 100);
+			var proxy = generator.CreateInterfaceProxyWithoutTarget<IOut>(new SetArgumentValueInterceptor(different));
+			proxy.Method(out x);
+
+			Assert.AreNotEqual(original.Value, x.Value);
+		}
+
+		[Test]
+		[TestCase(typeof(IByReadOnlyRef))]
+		public void By_reference_In_parameter_has_ParameterAttributes_In_set(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.True(parameter.Attributes.HasFlag(ParameterAttributes.In));
+		}
+
+		[Test]
+		[TestCase(typeof(IByValue))]
+		[TestCase(typeof(IByRef))]
+		[TestCase(typeof(IOut))]
+		public void Parameter_other_than_by_reference_In_does_not_have_ParameterAttributes_In_set(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.False(parameter.Attributes.HasFlag(ParameterAttributes.In));
+		}
+
+		[Test]
+		[TestCase(typeof(IByReadOnlyRef))]
+		public void By_reference_In_parameter_has_IsReadOnlyAttribute(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.True(parameter.GetCustomAttributes().Any(a => a.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute"));
+		}
+
+		[Test]
+		[TestCase(typeof(IByValue))]
+		[TestCase(typeof(IByRef))]
+		[TestCase(typeof(IOut))]
+		public void Parameter_other_than_by_reference_In_does_not_have_IsReadOnlyAttribute(Type type)
+		{
+			var parameter = type.GetMethod("Method").GetParameters()[0];
+
+			Assert.False(parameter.GetCustomAttributes().Any(a => a.GetType().FullName == "System.Runtime.CompilerServices.IsReadOnlyAttribute"));
+		}
+
+		public readonly struct ReadOnlyStruct
+		{
+			private readonly int value;
+
+			public ReadOnlyStruct(int value)
+			{
+				this.value = value;
+			}
+
+			public int Value => this.value;
+		}
+
+		public interface IByValue
+		{
+			void Method(ReadOnlyStruct arg);
+		}
+
+		public interface IByReadOnlyRef
+		{
+			void Method(in ReadOnlyStruct arg);
+		}
+
+		public interface IByRef
+		{
+			void Method(ref ReadOnlyStruct arg);
+		}
+
+		public interface IOut
+		{
+			void Method(out ReadOnlyStruct arg);
+		}
+
+		public sealed class SetArgumentValueInterceptor : IInterceptor
```

Comment:
This class should probably sit alongside `Castle.DynamicProxy.Tests.Interceptors.SetReturnValueInterceptor`, and also accept an index.

Reply:
Done.